Sming/Components/crypto/include/Crypto/HashContext.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <axtls-8266/crypto/crypto.h>
+#include <WString.h>
+#include <Data/HexString.h>
+
+namespace Crypto
+{
+/**
+ * @brief
+ */
+struct Blob {
+	const void* data;
+	size_t length;
+
+	Blob(const void* data, size_t length) : data(data), length(length)
+	{
+	}
+
+	Blob(const String& str) : data(str.c_str()), length(str.length())
+	{
+	}
+};
+
+/**
+ * @brief Class template for a Hash value, essentially just an array of bytes
+ */
+template <size_t size> struct HashValue {
+	uint8_t data[size];
+
+	static constexpr size_t length = size;
+
+	bool copy(const void* src)
+	{
+		if(src == nullptr) {
+			memset(data, 0, size);
+			return false;
+		}
+
+		memcpy(data, src, size);
+		return true;
+	}
+
+	bool operator==(const HashValue& other) const
+	{
+		return memcmp(data, other.data, sizeof(data)) == 0;
+	}
+
+	String toString(char separator = '\0') const
+	{
+		return makeHexString(data, size, separator);
+	}
+};
+
+/**
+ * @brief Class template for a Hash implementation 'Context'
+ */
+template <class Context, class Hash> class HashContext
```

Comment:
Looks like CRTP (`class Context` is the derived class). Isn't that overly complex? 

On first sight, it should be possible to avoid the inheritance altogether by making `Context` an "engine" class that is embedded as a private member inside HashContext.

After all, the goal is to create a common interface for different Hash algorithms, so all instantiations of HashContext should provide the same public interface.

Reply:
Took me a bit of time, but I get what you mean. Excellent idea. Thanks.